feat: add single node payment mode

mickvandijke opened the pull request:
Add an additional payment mode where only one node gets paid instead of three for each record. The one node will have to be paid three times the quoted amount as a penalty for using this mode. So uploads using the single node payment mode will be more expensive ANT token wise, but it will save gas fees by about 67%.

This mode is enabled by default. To disable this mode, simply add the `--disable-single-node-payment` flag to the `ant upload command`.

This is a client and smart contract only change.

REQUIREMENT:

- [x] Update the PROD `PaymentVault` smart contract to support single node payment mode.

mickvandijke committed:
feat: add single node payment mode

mickvandijke committed:
feat: add short flag `-s` for single node payment option

mickvandijke committed:
refactor(autonomi): fix clippy errors

mickvandijke committed:
fix(test): payment mode tests

mickvandijke committed:
fix(autonomi): refine single node payment logic and update comment

mickvandijke committed:
refactor(autonomi): extract and modularize payment processing logic

mickvandijke committed:
chore(ant): update single node payment message to clarify gas fees optimization

mickvandijke committed:
fix(test): update payment mode tests to multi-thread

maqi reviewed (changes requested):
Put as Request Change, as the new cmd opt of `single_node_payment` need a decision of `whether shall be default to be true`, i.e. whether the new scheme shall be enabled by default.

maqi commented on autonomi/src/client/quote.rs:
```diff
@@ -193,41 +203,172 @@ impl Client {
+    }
+
+    /// Create payment structure for single node mode (pay only highest priced node with 10x)
+    fn create_single_node_quote_payment(
```
the input quotes are already `sorted by price` 
the `create_standard_quote_payment` and `create_single_node_quote_payment` can simply to be:
```
        QuoteForAddress(vec![
            (*p1, a1.clone(), q1.clone(), Amount::ZERO),
            (*p2, a2.clone(), q2.clone(), Amount::ZERO),
            quotes[2].clone(),
            quotes[3].clone(),
            quotes[4].clone(),
        ])
```
or 
```
        QuoteForAddress(vec![
            (*p1, a1.clone(), q1.clone(), Amount::ZERO),
            (*p2, a2.clone(), q2.clone(), Amount::ZERO),
            (*p3, a3.clone(), q3.clone(), Amount::ZERO),
            (*p4, a4.clone(), q4.clone(), Amount::ZERO),
            quotes[4].clone(),
        ])
```

maqi commented on autonomi/src/client/quote.rs:
```diff
@@ -193,41 +203,172 @@ impl Client {
+            .max()
+            .expect("quotes should not be empty as we checked len >= MINIMUM_QUOTES_TO_PAY");
+
+        let node_to_pay = self.select_highest_priced_node(quotes, highest_price);
```
already sorted, no need to re-sort

mickvandijke replied:
Sharp!

maqi commented on autonomi/src/client/quote.rs:
```diff
@@ -193,41 +203,172 @@ impl Client {
+        let enhanced_price = highest_price * Amount::from(10u64);
+
+        // Sort by distance to maintain the closest 5 nodes
+        self.sort_quotes_by_distance(quotes, content_addr);
```
does this step required ?

mickvandijke replied:
No :)

mickvandijke committed:
refactor: change snp penalty from 10x to 3x

mickvandijke committed:
refactor(autonomi): simplify single node payment logic

mickvandijke committed:
fix(test): update payment mode tests

jacderida committed:
feat: use single-node payment by default

The decision was made to use single-node payment by default.

Therefore the `--single-node-payment` argument on the `file cost` and `file upload` commands is
inverted to `--disable-single-node-payment`.
